### PR TITLE
Add PROCESS.md documenting the new SDLC

### PR DESCRIPTION
**Pull request by jdesrosiers**
Resolves: https://github.com/json-schema-org/json-schema-spec/issues/1334

This is the formal documentation of the process we've discussed in https://github.com/orgs/json-schema-org/discussions/234. That discussion was focused on getting agreement on the general direction, while this makes some concrete decisions. Please do challenge any of the concrete decisions I've made here, but keep in mind that this is just a starting point, so it might be more productive address issues after merge as separate PRs/Issues so we don't end up with multiple threads of discussion in this PR.

Several more concrete decisions still need to be made that should be addressed in the spec (for example: dialect and vocabulary URIs). Those decisions are out of scope for this document.

I made some minor changes from what we had discussed to how `STAGE-2` is defined, so pay particular attention to that section when reviewing.

### CODE REVIEW

File: PROCESS.md
Comment (gregsdennis) [inline PROCESS.md:6]: Should we also say that it doesn't apply to vocabularies (like mine) or other extensions (like JSON Schema form generation) defined outside of the org?

File: PROCESS.md
Comment (gregsdennis) [inline PROCESS.md:29]: If we're considering flagless features as stable, is there another indicator as to when the feature reached stability?  Or does "doesn't have a flag" only pertain to the status flag?  Do we expect all stable features to have a year flag?  If one doesn't is that just an admin error that needs to be fixed?

Maybe it's easier to require that features have flags so that if they don't it's something to be fixed.  Having an allowance for not having a flag seems to raise a lot of questions.

File: PROCESS.md
Comment (gregsdennis) [inline PROCESS.md:52]: Grammar (maybe lose the second "for"?)

File: PROCESS.md
Comment (gregsdennis) [inline PROCESS.md:76]:
```suggestion
`STAGE-1` features are included in the specification, but flagged as `STAGE-1`
```

File: PROCESS.md
Comment (gregsdennis) [inline PROCESS.md:48]: I assume these checklists are for promotion _out of_ this stage to the next one?

File: PROCESS.md
Comment (gregsdennis) [inline PROCESS.md:83]:
```suggestion
feature experimentation and evolution should have been done in `STAGE-0`.
```

File: PROCESS.md
Comment (gregsdennis) [inline PROCESS.md:80]: I'm curious about how we will guage/measure this real world use.  Wouldn't we be reliant solely on volunteered feedback?

File: PROCESS.md
Comment (gregsdennis) [inline PROCESS.md:108]: > the vast majority

Do we need to define this more rigidly?

File: PROCESS.md
Comment (gregsdennis) [inline PROCESS.md:151]:
```suggestion
available as a PDF on the website at <placeholder>.
```

... if you intend to fill the blank later.

File: PROCESS.md
Comment (gregsdennis) [inline PROCESS.md:26]: Do we want to avoid contractions like we do in specs or do we think less formal language is okay for this doc?

**Review by gregsdennis — CHANGES_REQUESTED** [review]
We mention "JSON Schema Org core contributors" in several places.  Do we have/need a process to determine on- and offboarding of people into this role?

Also, a lot of my questions may be candidates for issues.  Feel free to indicate which ones you'd like to defer.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:6]: I thought I had covered that by explicitly stating what this process applies to (specs maintained by the JSON Schema Org), but technically it doesn't say that's the only thing it applies to. I'll edit to make that less ambiguous.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:26]: Are we sure using contractions in specs is discouraged? I don't think I've ever heard that. I think the language should be as natural as possible without being ambiguous. Using natural language wherever possible makes specs easier for readers to read and understand.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:29]: A part of the spec can only be unflagged if it's a stable feature that was introduced in the original stable release.

My thinking was that it would be too noisy to flag everything in the spec. It would be hard to see `2024` flags in a sea of `2023` flags. The first stable release wouldn't have any `2023` flags because it's unambiguous when those features were introduced since `2023` is the first release. Stable flags would start to be added for `2024`. That means that the number of flags in the spec should be fairly small and therefore stand out to readers.

I'll revise to clarify the purpose of unflagged features, but let's continue the discussion and get some more opinions if you still dislike the idea.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:48]: Correct. Any suggestions to make that more clear?

File: PROCESS.md
Comment (Julian) [inline PROCESS.md:159]:
```suggestion
Meta-schema URIs SHOULD NOT change once published, but if they do, the old URI
```

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:108]: Maybe? Assuming bowtie grows to include most implementations, this decision could be made based on hard data rather than gut feel or a time consuming audit of all known implementations.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:151]: Oops. Yeah, I had a placeholder originally and decided to remove it. I forgot to remove the "at" as well.

File: PROCESS.md
Comment (gregsdennis) [reply PROCESS.md:29]: Maybe we keep the "unflagged implies stable" idea, and instead of flagging the year for future features, we have a `newly-stable` flag that identifies features that are new for the year's release.  Then the next year, those `newly-stable` flags are cleared from those features so that the new `newly-stable` features stand out.

File: PROCESS.md
Comment (gregsdennis) [reply PROCESS.md:48]: Maybe explicitly say "Checklist for promotion to stage-X"

File: PROCESS.md
Comment (gregsdennis) [reply PROCESS.md:80]: I think in general, maybe.  But that may also be skewed toward more popular implementations that may or may not support a feature.

I like your answer below about including Bowtie and extracting metrics from there.  It shows support for a feature, but not necessarily usage.

File: PROCESS.md
Comment (gregsdennis) [reply PROCESS.md]: Your changes look fine.

File: PROCESS.md
Comment (gregsdennis) [reply PROCESS.md:30]: Just scanning the spec for `'s`s, it's predominantly possessives.  There are a couple `it's`s and `let's` (and even a `what's`), though.

I personally tend to steer clear of contractions in formal writing, but it doesn't appear that we as a group have been sufficiently consistent in our productions to say we have a preference.

File: PROCESS.md
Comment (gregsdennis) [reply PROCESS.md:88]: I don't think this is something that we have to resolve now.

**Review by gregsdennis — APPROVED** [review]
The changes that have been made are good.  I don't think the remaining items warrant holding this up.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md]: Let's discuss it at the next OCWM

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:88]: This is a tough one. I think it will always be largely a judgement call, but we should definitely be looking for more and better ways we can gather data we can use to make those calls.

**Review by Relequestual — CHANGES_REQUESTED** [review]
Left some comments and questions.

Some parts feel under specified, which is fine for a draft, but I feel like there are some key issues to iron out.

Additionally, I feel like we should be stating...
"This policy document uses RFC2119 defined key words to indicate requirement levels" =]

Bulk of this looks like a great first start though. Thanks for all your work on this!

**Comment by gregsdennis** [comment]
@Relequestual have a read through the comments that I made and the responses given. Maybe that will answer some of yours.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:38]: You quoted the answer immediately before asking the question, so I'm not sure how to answer.

> A feature can only be unflagged if it was stable in the initial stable release in 2023.

I can expand on that if it's not clear enough, but I'm not sure what's not clear. Does this discussion with @gregsdennis help? https://github.com/json-schema-org/json-schema-spec/pull/1348#discussion_r1025595948

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:45]: I fully expect champions to disappear on occasion. In that case, a new champion would be named or the feature would be abandoned and removed. The "SHOULD" is to give us some leeway if for some reason we can't or don't want to assign a champion right away, but also don't want to remove the feature. It's not a likely scenario. I'm fine with calling it a MUST if we want to.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:104]: Whoever the JSON Schema Org active contributors happen to be at the time. Maybe at some point that will be a specific and official group of people like a TSC, but for now it's squishy concept. For now I'll make the text more specific. We can update the wording when/if we have an official group we can name.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:105]: See, https://github.com/json-schema-org/json-schema-spec/pull/1348#discussion_r1028497441

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:129]: > "vast majority"

see https://github.com/json-schema-org/json-schema-spec/pull/1348#discussion_r1024614138

> "actively maintained"

Yep. The more I think about it, the more I agree that this needs more definition. For example, an implementation may be active for bug fixes, but don't intent to add new features. Should that count? This probably needs discussion and probably doesn't need to be decided now as it will be a while before we have to make this determination.

> "support the feature"

The meaning of those words seem pretty clear to me. We could explicitly say it has to pass all the tests. Is there more nuance that I'm missing?

> I think we should investigate further sponsoring implementation work at some point

Sure. I think that speaks to increasing the pool of "actively maintained" implementations.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:165]: Just like the spec, there is one set of mutable meta-schemas that describe the current state of the specification. Because of backward and forward compatibility constraints, no matter what release of the spec the schema is written for, the latest meta-schemas is all you need.

> Are we going to maintain meta-schemas attached to individual release snapshots?

We could have snapshots for reference like with the spec, but there wouldn't be reason to actually use those snapshots for anything other than historical reference.

> Are we going to have a "current" meta-schema which contains all keywords, even stage-0?
> Are we going to have a "current" meta-schema which per stage? For example, all stage-1 and above or all stage-2 and above?

The meta-schemas should include anything that's in the spec. That includes `STAGE-1`, `STAGE-2`, and deprecated.

> If we have a "current" meta-schema, which includes all of the stages, how can we update meta-schemas yet maintain backwards compatiblity?

Meta-schema changes should be backward compatible because spec changes are backward compatible.

> This becomes a very specific problem if we disallow unknown keywords by default.
>
> What if we wanted to update the meta-schemas to use a new keyword? That wouldn't be backwards compatible.

I was thinking that the standard dialect meta-schema would allow additional keywords. If a keyword appears in the meta-schema but is not supported by the implementation, the implementation should raise an error. It's not something the meta-schema can know, so I consider it a runtime constraint the implementation needs to deal with, not the meta-schema.

However, meta-schemas could disallow additional keywords and it shouldn't hurt anything. If an implementation is using an older version of the meta-schema that doesn't have a newer keyword used by the schema, then the implementation also wouldn't have support for the keyword. The result is always that you get an error that the implementation doesn't know the keyword.

Either way, there shouldn't be a backward or forward compatibility issue and whether we choose to allow additional keywords in the meta-schema is not relevant to process, so we can discuss that when we get to writing the meta-schemas.

Comment (jdesrosiers) [reply PROCESS.md:165]: I expect there to be some discussion about meta-schema lifecycle that may end up as a bigger conversation. This was the kind of thing I thought might be best to get into after the initial merge. Like I said, I don't expect merging this document to mean it's "done". I expect it to be a first draft that we will iterate on.

File: PROCESS.md
Comment (Relequestual) [reply PROCESS.md:45]: I'd rather call it a MUST. As for remaining in the spec, I think that would need to be a case by case basis.

File: PROCESS.md
Comment (Relequestual) [reply PROCESS.md:165]: There are a number of assertions in your response which I don't agree with, but I don't think here's the right place to capture this discussion.

>I expect there to be some discussion about meta-schema lifecycle that may end up as a bigger conversation. This was the kind of thing I thought might be best to get into after the initial merge. Like I said, I don't expect merging this document to mean it's "done". I expect it to be a first draft that we will iterate on.

I think we need to discuss how we can break up the SDLC into more discreet chunks for further discussion.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:38]: I see. I'll try to reword to avoid that misunderstanding. Removing or changing flags isn't disallowed, but compatibility rules apply. So, you couldn't remove a stable flag (such as `2024`) because it would be a backward compatible change, but you could change a `STAGE-2` to a `2024` because `STAGE-2` is allowed backward-incompatible changes. Similarly, you couldn't remove a `STAGE-2` flag without replacing it with a stable flag because that would break backward compatibility.

Comment (jdesrosiers) [reply PROCESS.md:38]: Here's an example of the kind of thing I have in mind from the node.js docs https://nodejs.org/dist/latest-v19.x/docs/api/fs.html. They have stable, unstable, and deprecated features and flag them inline just like we're talking about here.

File: PROCESS.md
Comment (jdesrosiers) [reply PROCESS.md:165]: That's exactly why I proposed merging this as a work-in-progress and opening issues to discuss the parts that need more discussion. Holding this up until we decide every detail makes it more difficult to have those discussions.

**Comment by Relequestual** [comment]
@jdesrosiers Can this be closed in favour of https://github.com/json-schema-org/json-schema-spec/pull/1500 ?

**Comment by jdesrosiers** [comment]
Yes, it can

**Pull request closed without merging**
